### PR TITLE
feat: Add classification metrics

### PR DESCRIPTION
**Pull request by josephmaa**
The aim of this PR is to provide more metrics besides accuracy that we can use for determining the effectiveness of a model. Currently adds auc roc curves.

### CODE REVIEW

**Comment by josephmaa** [comment]
Example from test dataset:

Screen Shot 2022-04-05 at 4 33 16 PM

**Comment by josephmaa** [comment]
Here are some of the results from printing the predicted outputs and the output of the decision function from the LinearSVC classifier:

```
[[0 0 0 0 0]
 [0 0 0 0 0]
 [1 0 0 0 0]
 [0 0 0 0 1]
 [0 0 0 0 1]
 [0 0 0 0 1]
 [1 0 0 0 0]
 [0 0 0 0 0]
 [0 0 0 0 0]
 [0 0 0 1 0]
 [0 0 0 1 0]
 [0 1 0 0 0]
 [0 1 0 0 0]
 [1 0 0 0 0]
 [0 0 0 0 1]
 [0 0 0 0 1]
 [0 1 0 0 0]]
[[-0.6488396  -0.96458057 -0.05070014 -0.33378502 -0.96120859]
 [-0.6488396  -0.96458057 -0.05070014 -0.33378502 -0.96120859]
 [ 0.91559418 -0.97292689 -0.97992985 -0.96112776 -0.97033759]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [ 0.91559418 -0.97292689 -0.97992985 -0.96112776 -0.97033759]
 [-0.6488396  -0.96458057 -0.05070014 -0.33378502 -0.96120859]
 [-0.6488396  -0.96458057 -0.05070014 -0.33378502 -0.96120859]
 [-0.95654125 -0.96457759 -0.97377657  0.8969803  -0.96121106]
 [-0.95654125 -0.96457759 -0.97377657  0.8969803  -0.96121106]
 [-0.9623366   0.63068604 -0.97727329 -0.6892753  -0.96637851]
 [-0.9623366   0.63068604 -0.97727329 -0.6892753  -0.96637851]
 [ 0.91559418 -0.97292689 -0.97992985 -0.96112776 -0.97033759]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [-0.9623366   0.63068604 -0.97727329 -0.6892753  -0.96637851]]
```

**Comment by josephmaa** [comment]
Unexpectedly, the fit using the prediction probabilities from the pyuoi `predict_proba` function yielded high values for the area under the roc-auc curves. (see below)

Screen Shot 2022-04-09 at 1 02 15 PM
Screen Shot 2022-04-09 at 1 02 36 PM
Screen Shot 2022-04-09 at 1 02 40 PM
Screen Shot 2022-04-09 at 1 02 43 PM
Screen Shot 2022-04-09 at 1 02 46 PM

**Comment by josephmaa** [comment]
Have generated the precision-recall curves with the LinearSVC model. I also increased the number of samples since the first dataset didn't have enough examples of the rarer events to generate useful precision-recall curves. 

Screen Shot 2022-04-09 at 1 30 16 PM
Screen Shot 2022-04-09 at 1 30 28 PM
Screen Shot 2022-04-09 at 1 30 32 PM

**Comment by shwan19** [comment]
A few questions:
1) What was in the input dataset you used for the linear SVC? (I'm guessing not the behavior data?)
2) How many positive examples are there of class 1 v. class 2 v. class 3 in the PR plots?
3) Why are there 4 classes for ROC plots and 3 classes for PR plots?

But I think there is a straightforward answer to your question about the class 3 PR plot. 1 edge v. more than 1 edge is not the central point, I don't think. Take a look at the y-axes for class 1 v. class 2 v. class 3. You'll notice that the lowest value on the y-axis is very quite different across these three plots. The remaining portion of the plot (between y = 0 and y = min) is implied/hidden (though of course you can implement plotting differently to get the whole plot). The AUPRC is thus the area between each PRC and y = 0 for x = [0,1].

**Comment by josephmaa** [comment]
> A few questions:
> 
> 1. What was in the input dataset you used for the linear SVC? (I'm guessing not the behavior data?)

The input dataset was generated with the `--use_small_dataset` flag, so not the behavior data

```
if parsed_args.use_small_dataset:
        n_features = 20
        n_inf = 10
        x, y, _, _ = make_classification(n_samples=1000,
                                         random_state=10,
                                         n_classes=5,
                                         n_informative=n_inf,
                                         n_features=n_features,
                                         shared_support=True,
                                         w_scale=4.)

        x_train, x_test, y_train, y_test = train_test_split(
            x, y, random_state=parsed_args.training_seed)
```

> 2. How many positive examples are there of class 1 v. class 2 v. class 3 in the PR plots?

Here's the confusion matrix of values:

Screen Shot 2022-04-10 at 2 07 22 AM

I added some code so that we can visualize the results for all subsequent runs.

> 3. Why are there 4 classes for ROC plots and 3 classes for PR plots?

That's a good catch. There was a bug in the code that I think was related to the ordering of the plots getting generated. The final plot is here. You can see in the commit after this comment that I changed the code to plot the precision-recall curves first.

> But I think there is a straightforward answer to your question about the class 3 PR plot. 1 edge v. more than 1 edge is not the central point, I don't think. Take a look at the y-axes for class 1 v. class 2 v. class 3. You'll notice that the lowest value on the y-axis is very quite different across these three plots. The remaining portion of the plot (between y = 0 and y = min) is implied/hidden (though of course you can implement plotting differently to get the whole plot). The AUPRC is thus the area between each PRC and y = 0 for x = [0,1].

That is a good point. I think some articles mention plotting the no-skill line "The no-skill line changes based on the distribution of the positive to negative classes. It is a horizontal line with the value of the ratio of positive cases in the dataset.". If we assume the no-skill line to pick a random class, since the dataset is fairly evenly distributed, I would expect the line to be plotted at approx. 0.2 on the y-axis for these precision-recall curves. I'm still thinking about how we would have a no-skill line based on the classifier picking only one class (the alternative[ mentioned in the article](https://machinelearningmastery.com/roc-curves-and-precision-recall-curves-for-classification-in-python/#:~:text=A%20precision%2Drecall%20curve%20is,constant%20class%20in%20all%20cases.)).

File: classifier/roc_curve.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    expected = [1 if e == label else 0 for e in expected]
+    observed = [1 if o == label else 0 for o in expected]
+    df['expected'] = expected
+    df['observed'] = observed
+
+    # Sort based on confidence values
+    df.sort_values('confidence', ascending=False)
+
+    # Accumulate values and calculate tpr and fpr rates per element
+    df['tpr'] = 0
```
Comment (kdquine) [inline classifier/roc_curve.py:36]: this looks unfinished

Comment (josephmaa) [reply classifier/roc_curve.py]: removed

**Comment by josephmaa** [comment]
Did some testing with the `auc` function to test its functionality on multi-class classification since that was one of the questions that we had with it. Using the values from [the documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.auc.html), I tested alternative use of binary labeling prior to getting passed into `fpr` and `tpr` which worked well.

Screen Shot 2022-07-12 at 2 40 44 PM

**Comment by josephmaa** [comment]
Here are some real-data from a single class within a PC:
Screen Shot 2022-07-12 at 4 44 49 PM
Screen Shot 2022-07-12 at 4 44 43 PM
Screen Shot 2022-07-12 at 4 44 37 PM
Screen Shot 2022-07-12 at 4 44 33 PM
Screen Shot 2022-07-12 at 4 44 28 PM
